### PR TITLE
Nomos node tree overlay

### PR DESCRIPTION
**Pull request by bacv**
Added tree overlay to nomos-node. Some configuration parameters were missing in example config.yaml, so the file was updated.

### CODE REVIEW

File: mixnet/client/src/config.rs [hunk 1]
```diff
@@ -11,7 +11,9 @@ pub struct MixnetClientConfig {
     pub mode: MixnetClientMode,
     pub topology: MixnetTopology,
     pub connection_pool_size: usize,
+    #[serde(default = "MixnetClientConfig::default_max_retries")]
     pub max_retries: usize,
+    #[serde(default = "MixnetClientConfig::default_retry_delay")]
```
Comment (danielSanchezQ) [inline mixnet/client/src/config.rs:16]: Is this from another branch?

File: nodes/nomos-node/config.yaml [hunk 2]
```diff
@@ -49,3 +53,12 @@ http:
...
     cors_origins: []
+
+da:
+  da_protocol:
+    num_attestations: 1
+  backend:
+    max_capacity: 10
+    evicting_period: 
+      secs: 3600
+      nanos: 0
```
Comment (danielSanchezQ) [inline nodes/nomos-node/config.yaml:64]: Same, this is not from this pr or is it?

**Review by danielSanchezQ — APPROVED** [review]
Some of the overlay config will have to go away eventually as it must be calculated from chain instead. Overall looks good. I just do not know we we didn't mix a couple branches in this PR.

**Review by al8n — APPROVED** [review]

File: mixnet/client/src/config.rs (hunk 1 above)
Comment (bacv) [reply mixnet/client/src/config.rs:16]: No, the another branch had defaults for `MixnetNode`, I've added these for `MixnetClient` config parsing.

File: nodes/nomos-node/config.yaml (hunk 2 above)
Comment (bacv) [reply nodes/nomos-node/config.yaml:64]: This is from this PR, had to add these changes for config to work.
Added these changes to PR because they are small, these are separate commits:
image

**Review by youngjoon-lee — APPROVED** [review]

**Comment by bacv** [comment]
@Zeegomo - After changing FlatOverlay to TreeOverlay in nomos node and running integration tests, ten happy path nodes network had some panicking nodes and failed to progress with views. @danielSanchezQ helped debug it and error handling for this case was added. Please review. 

I've also increased the unhappy case timeout, tree overlay added more latency.

File: nomos-services/consensus/src/lib.rs [hunk 3]
```diff
@@ -229,7 +229,8 @@ where
                 )
                 .await
                 else {
-                    unreachable!()
+                    tracing::debug!("Failed to gather initial votes");
+                    return Event::None;
```
Comment (danielSanchezQ) [inline nomos-services/consensus/src/lib.rs:233]: @Zeegomo This should be ok right? I have been working with Gusto and in reality if failing to gather the initial votes nodes should catch up anyway.

**Comment by Zeegomo** [comment]
I'll review it after our meeting 👍🏼

File: nomos-services/consensus/src/lib.rs (hunk 3 above)
Comment (Zeegomo) [reply nomos-services/consensus/src/lib.rs:233]: Failing to gather initial votes means we go to the timeout routine, why does this happen now?

**Review by Zeegomo — COMMENTED** [review]
What were the panics you saw in tests? I suspect they were caused by a timeout triggering the closing of the vote stream.
Those changes seems ok (and we were looking to add them), but since the tree overlay with 1 committee is exactly like a flat one, I'd like to investigate why we need the additional time

**Comment by danielSanchezQ** [comment]
> What were the panics you saw in tests? I suspect they were caused by a timeout triggering the closing of the vote stream. Those changes seems ok (and we were looking to add them), but since the tree overlay with 1 committee is exactly like a flat one, I'd like to investigate why we need the additional time
> 
> > I've also increased the unhappy case timeout, tree overlay added more latency.
> 
> As in, this _shouldn't_ be the case for a 1 committee tree overlay

Yes, issue need to be investigated. Apparantly after some number of nodes, some get unresponsive 🤔 , starting with the first leader.

**Comment by Zeegomo** [comment]
I think it's worth investigating this issue now that the playground is still small, debugging something like this at greater scales is almost impossible

**Comment by bacv** [comment]
> What were the panics you saw in tests? I suspect they were caused by a timeout triggering the closing of the vote stream.
> Those changes seems ok (and we were looking to add them), but since the tree overlay with 1 committee is exactly like a flat one, I'd like to investigate why we need the additional time

The panic was caused by executing `unreachable` code here: https://github.com/logos-co/nomos-node/pull/415/files#diff-12bf2e7f638b9635b046010a05a3ac31b6c7abf580576c6a6b04bca1f33a394fL48-L83

It's most likely caused by the tree overlay. The increase in the time it takes to complete the test cases is probably related to the added error handling that triggers unhappy path?

> I think it's worth investigating this issue now that the playground is still small, debugging something like this at greater scales is almost impossible

Agree, I'll ping you how we could organize this.

**Comment by bacv** [comment]
Moved code unrelated to tree overlay to #423

File: tests/src/nodes/nomos.rs [hunk 4]
```diff
@@ -263,13 +263,12 @@ fn create_node_config(
...
+            overlay_settings: TreeOverlaySettings {
                 nodes,
                 leader: RoundRobin::new(),
-                // By setting the leader_threshold to 1 we ensure that all nodes come
-                // online before progressing. This is only necessary until we add a way
-                // to recover poast blocks from other nodes.
-                leader_super_majority_threshold: Some(threshold),
+                current_leader: [0; 32].into(),
+                number_of_committees: 1,
+                committee_membership: RandomBeaconState::initial_sad_from_entropy([0; 32]),
```
Comment (youngjoon-lee) [inline tests/src/nodes/nomos.rs:271]: I've tested the `ten_nodes_happy` with disabling mixnet and revert the `unreachable!()`, in order to check the reason why local_timeout occurs in the happy path test. (I reverted the `unreachable!()` because the stream is closed when local_timeout occurs in our case).

**In short, local_timeout occurs due to the design of our test environment, only when we use tree-overlay with random-beacon.**

For example, there are 10 nodes (node-0~9). Let’s say the next leader (of view 1) is node-6, selected by random-beacon. In our test environment, 10 nodes are started sequentially (from 0 to 9). In other words, node-6 is not started until node-5 becomes ready.

In this case, I saw that node-6 received only 4 votes for view 0, and all of those votes were from node-6\~9 which were started after node-6. **That is, votes from node-0~5 (started before node-6) weren’t delivered to the node-6.**

In fact, this is the issue that we’re already aware of. Our current libp2p network backend doesn’t provide a way to catch-up past messages. 

Btw, this panic didn’t happen in the previous test with flat-overlay because the round-robin is used instead of random-beacon. That is, the next leader (of view 1) is node-0 which is started at first. So, this is no problem for node-0 to receive votes from all other nodes that are started after itself.

To make the test work with tree-overlay + random-beacon, we may need to pause the consensus service of all nodes until all 10 nodes are started successfully. Actually, we have a 1s `sleep` before starting the consensus service: https://github.com/logos-co/nomos-node/pull/338, but this might not be enough. I'm thinking about other solutions.

Comment (danielSanchezQ) [reply tests/src/nodes/nomos.rs:271]: Nice job!! all this seems about right. Now we need to tackle it so everything is clear next time.

Comment (youngjoon-lee) [reply tests/src/nodes/nomos.rs:271]: Next steps discussed:
- Modify the test to check if all blocks are happy block
- Modify the test to start the node first who is going to be the next leader (instead of always starting the node-0 first)

Comment (youngjoon-lee) [reply tests/src/nodes/nomos.rs:271]: Ah.. I just found that we're using both `RoundRobin` and `RandomBeaconState`. So, let me rephrase the sentence above to "We use `RoundRobin` for `LeaderSelection` but the node list is suffled at the beginning (and whenever overlay is refreshed) by `CommitteeMembership` which is `RandomBeacon`."

**Comment by youngjoon-lee** [comment]
I merged https://github.com/logos-co/nomos-node/pull/425 to this branch, but still we need https://github.com/logos-co/nomos-node/pull/426 to be merged to make the ten-nodes-happy test passed.

File: tests/src/nodes/nomos.rs (hunk 4 above)
Comment (youngjoon-lee) [reply tests/src/nodes/nomos.rs:258]: Resolved by https://github.com/logos-co/nomos-node/pull/425

**Comment by youngjoon-lee** [comment]
I merged all fixes to this PR. If CI passes, I think we can merge this.
